gufe/storage/externalresource.py

```diff
@@ -0,0 +1,113 @@
+import hashlib
+import warnings
+import pathlib
+
+from typing import ClassVar
+
+
+class ExternalResourceError(Exception):
+    """Base class for errors due to problems with external resources"""
+    # TODO: is it necessary to have a base class here? Would you ever have
+    # one catch that handles both subclass errors?
+
+
+class MissingExternalResourceError(ExternalResourceError):
+    """Error when the external resource could not be loaded"""
+
+
+class ChangedExternalResourceError(ExternalResourceError):
+    """Error when there's a SHA256 mismatch with an external resource"""
+
+
+class ExternalStorage:
+    """Abstract base for external storage."""
+
+    allow_changed: ClassVar[bool] = False
+
+    def validate(self, location, sha2):
+        if not self.get_sha2(location) == sha2:
+            msg = (f"Hash mismatch for {location}: this object "
+                   "may have changed.")
+            if not self.allow_changed:
+                # NOTE: having it here instead of in a DataLoader means that
+                # you can only change it for the whole system, instead of
+                # for a specific object
+                raise ChangedExternalResourceError(
+                    msg + " To allow this, set ExternalStorage."
+                    "allow_changed = True"
+                )
+            else:
+                warnings.warn(msg)
+
+    def get_sha2(self, location: str):
+        filelike = self._as_filelike(location, bytes_mode=True)
+        sha2 = hashlib.sha256()
+        # TODO: this can be probably be improved by chunking
+        sha2.update(filelike.read())
```

Comment:
Probably a future thing, but would it be worth also storing last modified time as a quicker hash, then actually compute sha if that fails?

Reply:
Yep (both on reasonable possibility, and on hold for future). Anticipating this, I changed the method name to `get_metadata`. Currently only returns the hexdigest string, but in the future that might be a (named?) tuple of metadata.

There may be some backend-specific weirdness (like if some backend doesn't preserve last-modified) but this should be backend-specific, and we could use `None` as a sentinel value for that.

Reply:
On this particular point: Is it worth wrapping the metadata in that namedtuple already? Right now it would only have one field (`md5`), but it would be easier to extend it in the future (in ways that might be more easily backward compatible).

The argument against it is YAGNI. Keep in mind that other metadata (last modified, etc.) will probably only be used as a fast-failure approach. A successful check will still need to calculate the md5. And I'd expect that most of the time, the file passes the metadata check. So adding paths for fast failure doesn't seem like something we should spend significant time on.